src/nvim/ui.c

```diff
@@ -182,6 +183,37 @@ static void ui_refresh_event(void **argv)
   ui_refresh();
 }
 
+void ui_notify_changed_highlights(garray_T changed_highlights)
+{
+  // we might want to send batchs instead
+  // + save current hl_id value to restore it later
+  // for backwards compatibility
+  ILOG("%d changed highlights", changed_highlights.ga_len);
+  Array args = ARRAY_DICT_INIT;
+  Error err;
+  bool refresh_cursor = false;
+  for (int idx = 0; idx < changed_highlights.ga_len; idx++) {
+    int hl_id = ((int *)changed_highlights.ga_data)[idx];
+    Dictionary dic = nvim_hl_from_id(hl_id, &err);
+    ADD(args, DICTIONARY_OBJ(dic));
+    // ILOG("comparing with %d", shape_table[cursor_get_mode_idx()].id);
+    if (hl_id == shape_table[cursor_get_mode_idx()].id) {
+      refresh_cursor = true;
+    }
+  }
+
+  if (args.size) {
+    ILOG("submit %d updated hl", args.size);
+    ui_event("highlights", (args));
+  }
+
+  if (refresh_cursor) {
+    ILOG("updating cursor !");
+    Array temp = ARRAY_DICT_INIT;
+    ui_event("refresh_cursor", temp);
```

Comment:
ui_cursor_shape() tries to be smart and doesn't send the event if the mode hasn't changed since last update. I would need to add a "force_update" boolean to bypass that behavior. Would that be ok ?